pkg/logging/journald_core.go

```diff
@@ -0,0 +1,122 @@
+// loggingFatal is set to true when fatals are logged so that other non-fatals can be swallowed.
+// You can find an example usage under Logging.loggingFatal.
+func NewJournaldCore(identifier string, loggingFatal *atomic.Bool, enab zapcore.LevelEnabler) zapcore.Core {
+	return &journaldCore{
+		LevelEnabler: enab,
+		identifier:   identifier,
+		loggingFatal: loggingFatal,
+	}
+}
+
+type journaldCore struct {
+	zapcore.LevelEnabler
+	context      []zapcore.Field
+	identifier   string
+	loggingFatal *atomic.Bool
+}
+
+func (j *journaldCore) Check(ent zapcore.Entry, ce *zapcore.CheckedEntry) *zapcore.CheckedEntry {
+	if j.Enabled(ent.Level) {
+		return ce.AddCore(ent, j)
+	}
+
+	return ce
+}
+
+func (j *journaldCore) Sync() error {
+	return nil
+}
+
+func (j *journaldCore) With(fields []zapcore.Field) zapcore.Core {
+	return &journaldCore{
+		LevelEnabler: j.LevelEnabler,
+		context:      append(j.context[:len(j.context):len(j.context)], fields...),
+		identifier:   j.identifier,
+		loggingFatal: j.loggingFatal,
+	}
+}
+
+// Write sends the log entry to systemd-journald.
+//
+// If fatal entries are logged, Write sleeps for a short time after the message has been written,
+// before zap then exits the program,
+// as systemd may not correctly attribute all or part of the log message due to a race condition,
+// and therefore it would be missing when checking logs via journalctl -u icingadb:
+// https://github.com/systemd/systemd/issues/2913
+//
+// Without sleep, we observed the following exit behavior when starting Icinga DB via systemd:
+//
+// • All or part of the log message is lost,
+// regardless of whether it is logged to stderr or to systemd-journald.
+//
+// • When writing to stderr after also writing to systemd-journald.
+// the message seems to be attributed correctly,
+// which is good for panics not under our control as they print to stderr.
+// For messages under our control,
+// this is actually not an option,
+// as we only want to log to systemd-journald.
+//
+// • Sleep before exiting helps systemd to attribute the message correctly,
+// regardless of whether it is written to systemd-journald or stderr.
+func (j *journaldCore) Write(ent zapcore.Entry, fields []zapcore.Field) error {
+	all := make([]zapcore.Field, 0, len(fields)+len(j.context))
+	all = append(all, j.context...)
+	all = append(all, fields...)
+
+	if ent.Level == zapcore.FatalLevel {
+		// Sleep between fatal log and os.Exit() to overcome a race condition that
+		// results in systemd no longer attributing messages from processes
+		// that have exited to their cgroup.
+		// See: https://github.com/systemd/systemd/issues/2913.
+		defer time.Sleep(time.Millisecond * 1100)
+
+		j.loggingFatal.Store(true)
+	}
+
+	pri, ok := priorities[ent.Level]
+	if !ok {
+		return errors.Errorf("unknown log level %q", ent.Level)
+	}
+
+	journalFields := make(map[string]string, len(all)+1)
+	journalFields["SYSLOG_IDENTIFIER"] = ent.LoggerName
```

Comment:
This includes the child logger name resulting in messages like `icingadb.high-availability[4129715]: Removed 0 old instances`. Typically that identifier is just the program name. This also made me miss most log messages at first as I used `journalctl -t icingadb`, which still finds some messages, so that filter looks good at first, but is confusing as messages are missing that you'd expect to be there.

Reply:
True. I wanted to include the child logger name and thought that it's also nice to be able to filter for certain components directly via `-t`. You would have `-u` for all the logs. Anyway, should it be `icingadb [4129715]: high-availability: Removed ...`?